lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,222 @@
+      # of the same class can be written using this method. The following
+      # failure scenarios are prevented before being sent to DynamoDB:
+      #
+      #   - One or more tables specified in the BatchWriteItem request does not
+      #      exist.
+      #   - You try to perform multiple operations on the same item in the same
+      #      BatchWriteItem request. For example, you cannot put and delete the
+      #      same item in the same BatchWriteItem request.
+      #   - There are more than 25 requests in the batch.
+      #   - Primary key attributes specified on an item in the request do not
+      #      match those in the corresponding table's primary key schema.
+      #
+      # This method will not prevent the following failure scenarios and will
+      # need to be handled separately:
+      #
+      #   - Any individual item in a batch exceeds 400 KB.
+      #   - The total request size exceeds 16 MB.
+      #
+      # If any items are not successfully submitted, a RuntimeError is raised
+      # and the BatchWriter can be inspected to see what the errors were.
+      # A call to #errors will return an array of response objects that contain
+      # the errors. See the documentation for Response at
+      # http://docs.aws.amazon.com/sdkforruby/api/Seahorse/Client/Response.html
+      #
+      # @param [Hash] opts options to pass on to the client call to
+      #   +#batch_item_write+. See the documentation above in the AWS SDK for
+      #   Ruby V2.
+      # @opts retry_count the number of retries to attempt on failed write
+      # @raise RuntimeError if unsuccessful in submitting all items
+      def save!(opts = {})
+        raise "can't submit if writer in #{@state} state." unless pending?
+        validate_items
+        # the -1 is to make it equal the number of retries requested
+        @retry_times = (opts.delete(:retry_count) { |_| 4 } - 1)
+        retry_items = send_chunks(opts)
+        responses = retry_failed_requests(retry_items, opts)
+        analyze_result(responses)
+      end
+
+      # Perform the save! function but capture the error that is raised on
+      # unsuccessful save.
+      def save(opts = {})
+        save!(opts)
+        true
+      rescue RuntimeError
+        false
+      end
+
+      def error?
+        @state == ERROR_SEND_STATE
+      end
+
+      # Get a hash of the responses that didn't work along with their
+      # unprocessed items. Mimics ActiveRecord design.
+      def errors
+        @errors.map do |err|
+          {
+            error: err.error,
+            data: err.data,
+            unprocesssed_items: err.unprocessed_items
+          }
+        end
+      end
+
+      def pending?
+        @state == PENDING_STATE
+      end
+
+      def unprocessed_items
+        # Will only have one error if error is RecordError
+        return [] if @errors.empty? || @errors.first.data.is_a?(Errors::RecordError)
```

Comment:
This is not necessarily semantically correct. For example, a call to `#batch_write_items` could return a throughput error because it wrote no items. It won't have a return value with unprocessed items, but we can correctly surmise that no items were processed. Returning an empty array presumes the opposite.

This would also be a reason to not squash the throughput exception from the client.